pkg/detectors/uri/uri.go

```diff
@@ -23,7 +23,7 @@ var _ detectors.Detector = (*Scanner)(nil)
 var _ detectors.CustomFalsePositiveChecker = (*Scanner)(nil)
 
 var (
-	keyPat = regexp.MustCompile(`\b(?:https?:)?\/\/[\S]{3,50}:([\S]{3,50})@[-.%\w\/:]+\b`)
+	keyPat = regexp.MustCompile(`\b(?:https?:\/\/)?[\w-\.$~!]{3,50}:([\w-\.%$^&#]{3,50})@[-.\w]+\b`)
```

Comment:
This is missing a large number of valid characters for usernames and passwords. The host pattern is also still fairly permissive and would match things that could never be valid, e.g. `@----__-2as-2`.

Reply:
How about something like:
```regex
\b(?:https?:\/\/)?[\w-\.$~!&'()*+,;=:%-]{3,50}:([\w-\.%$^#&'()*+,;=:%-]{3,50})@[a-zA-Z0-9.-]+(?:\.[a-zA-Z]{2,})?\b
```
Added additional valid characters and fixed the host pattern too.

Reply:
Example: https://regex101.com/r/qBF4nS/1

Reply:
That's looking a bit better. Some notes:

1. The scheme prefix shouldn't be optional
2. `:` isn't a valid username character
3. Username isn't always required ([example](https://stackoverflow.com/a/68783135))
4. The username and password patterns are both missing special characters. It would be pragmatic to add [all applicable special characters from `[[:graph:]]`](https://pkg.go.dev/regexp/syntax), and remove them later if they're causing issues. e.g.,
  ```regexp
  \bhttps?:\/\/[\w!#$%&()*+,\-./;<=>?@[\\\]^_{|}~]{0,50}:([\w!#$%&()*+,\-./:;<=>?[\\\]^_{|}~]{3,50})@[a-zA-Z0-9.-]+(?:\.[a-zA-Z]{2,})?\b
  ```
5. The pattern needs to be able to detect port as well as path

Reply:
Updated with port detection